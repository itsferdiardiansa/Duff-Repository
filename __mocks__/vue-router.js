global.mockRouterPush = jest.fn();
global.mockRouterAddRoute = jest.fn();
global.mockRouterBeforeEach = jest.fn();
global.mockRouterAfterEach = jest.fn();
global.mockRouterOnError = jest.fn();

// useRoute
global.mockRoutePath = jest.fn();
global.mockRouteFullPath = jest.fn();

module.exports = {
  ...jest.requireActual('vue-router'),
  createWebHistory: jest.fn(),
  createRouter: jest.fn(),
  useRouter: () => ({
    push: mockRouterPush,
    addRoute: mockRouterAddRoute,
    beforeEach: mockRouterBeforeEach,
    afterEach: mockRouterAfterEach,
    onError: mockRouterOnError,
  }),

  useRoute: () => ({
    path: mockRoutePath,
  }),
};
