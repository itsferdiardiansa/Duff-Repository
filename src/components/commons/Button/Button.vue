<template>
  <button
    :class="customClass"
    :disabled="disabled"
    :title="title"
  >
    <div :class="`${prefixClass}-button--wrapper`">
      <template v-if="icon.length">
        <div :class="`${prefixClass}-button--icon`">
          <font-awesome-icon :icon="icon" />
        </div>
      </template>

      <template v-if="hasSlot || label.length">
        <div :class="[`${prefixClass}-button--label`, {'has-icon': icon.length}]">
          <slot>
            {{ label }}
          </slot>
        </div>
      </template>
    </div>
  </button>
</template>
<script>
import { computed, getCurrentInstance } from 'vue'

export default {
  props: {
    label: {
      type: String,
      default: '',
    },
    title: {
      type: String,
      default: '',
    },
    tooltip: {
      type: String,
      default: ''
    },
    inverse: {
      type: Boolean,
      default: false
    },
    variant: {
      type: String,
      default: 'light',
      validator: function(variant) {
        return ~['primary', 'danger', 'warning', 'dark', 'success', 'light'].indexOf(variant)
      }
    },
    size: {
      type: String,
      default: 'base',
      validator: function(variant) {
        return ~['xs', 'sm', 'base', 'lg', 'xl'].indexOf(variant)
      }
    },
    icon: {
      type: Array,
      default: () => []
    },
    bold: {
      type: Boolean,
      default: false,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    rounded: {
      type: Boolean,
      default: true,
    },
    pill: {
      type: Boolean,
      default: false
    }
  },
  setup(props, { emit, slots }) {
    let root = getCurrentInstance()

    const getVariantClass = (prefixBtnClass) => {
      const { inverse, variant } = props 
      let btnClass = `${prefixBtnClass}--${variant}`

      if(inverse) btnClass += '-inverse'
      
      return btnClass
    }

    const customClass = computed(() => {
      const { data: { prefixClass } } = root
      const { size, disabled, rounded, bold, pill } = props
      const btnClass = prefixClass.concat('-button')
      let _class = [btnClass]

      _class.push(getVariantClass(btnClass))

      if(size) _class.push(btnClass.concat('--' + size))

      if (bold) _class.push('text-bold')

      if(pill)
        _class.push('pill')
      else
        if (!rounded) _class.push('no-rounded')

      if (disabled) _class.push('disabled')

      return [_class.join(' ')]
    })

    const handleClick = e => {
      emit('click', e)
    }

    const hasSlot = computed(() => (
      !!slots.default
    ))

    return {
      customClass,
      handleClick,
      hasSlot
    }
  },
}
</script>
<style lang="scss">
@import './variant.scss';
@import './size.scss';

.#{$prefixClass}-button {
  @apply h-9 focus:outline-none rounded px-3 transition duration-300 relative;
  @apply hover:bg-gray-300 text-gray-800 overflow-hidden whitespace-nowrap;

  @include variant;

  @include size;

  &.no-rounded {
    @apply rounded-none;
  }

  &.pill {
    @apply rounded-full;
  }

  &.text-bold {
    @apply font-bold;
  }

  &.disabled {
    @apply disabled:opacity-70 pointer-events-none;
  }
  
  &--wrapper {
    @apply flex-grow-0 flex flex-wrap items-center justify-center;
  }

  &--label {
    @apply pt-0.5;

    &.has-icon {
      @apply ml-2;
    }
  }
}
</style>
