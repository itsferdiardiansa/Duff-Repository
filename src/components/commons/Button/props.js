export default {
  label: {
    type: String,
    default: '',
  },
  inverse: {
    type: Boolean,
    default: false
  },
  variant: {
    type: String,
    default: 'light',
    validator: function(variant) {
      return ~['primary', 'danger', 'warning', 'dark', 'success', 'light'].indexOf(variant)
    }
  },
  size: {
    type: String,
    default: 'base',
    validator: function(variant) {
      return ~['xs', 'sm', 'base', 'lg', 'xl'].indexOf(variant)
    }
  },
  icon: {
    type: Array,
    default: () => []
  },
  bold: {
    type: Boolean,
    default: false,
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  rounded: {
    type: Boolean,
    default: true,
  },
  pill: {
    type: Boolean,
    default: false
  },
}