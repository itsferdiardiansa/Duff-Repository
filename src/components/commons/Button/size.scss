@mixin size {
  &.xs {
    @apply text-xs py-1.5 px-2;
  }

  &.sm {
    @apply text-sm p-2;
  }

  &.base {
    @apply text-base p-3;
  }

  &.lg {
    @apply text-lg p-4;
  }
}