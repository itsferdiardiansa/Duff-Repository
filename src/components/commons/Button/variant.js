export const buttonSize = {
  sm: 'py-1 px-2',
}