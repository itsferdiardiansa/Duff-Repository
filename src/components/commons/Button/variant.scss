@mixin variant {
  &.default {
    @apply bg-gray-200 hover:bg-gray-300 text-gray-800;
  }

  &.primary {
    @apply bg-blue-700 hover:bg-blue-800 text-gray-200;
  }

  &.danger {
    @apply bg-red-700 hover:bg-red-800 text-gray-200;
  }

  &.warning {
    @apply bg-yellow-500 hover:bg-yellow-600;
  }

  &.success {
    @apply bg-green-500 hover:bg-green-600 text-gray-200;
  }

  &.dark {
    @apply bg-gray-700 hover:bg-gray-800 text-gray-200;
  }
}