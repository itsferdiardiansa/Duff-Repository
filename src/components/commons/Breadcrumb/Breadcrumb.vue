<template>
  <ul :class="`${prefixClass}-breadcrumb`" ref="breadcrumbEl">
    <template v-for="(item, key) in data" :key="key">
      <li
        :class="`${prefixClass}-breadcrumb--item`"
        :ref="collectBreadcrumItemEl"
      >
        <template v-if="data.length - 1 === key">
          <Button
            size="sm"
            variant="orange"
            :label="item.name"
            :inverse="true"
            :rounded="false"
            :disabled="true"
          />
        </template>
        <template v-else>
          <Button
            size="sm"
            variant="orange"
            :rounded="false"
            :label="item.name"
            @click="changePage(item.path)"
          />
        </template>
      </li>
    </template>
  </ul>
</template>
<script>
/* eslint-disable */
import { ref } from 'vue';
import { useRouter } from 'vue-router';
import Button from '@common/Button';

export default {
  name: 'CMBreadcrumb',
  components: {
    Button,
  },
  props: {
    data: {
      type: Array,
      default: () => [],
    },
    separator: {
      type: String,
      default: '/',
    },
  },
  setup(props) {
    const router = useRouter();
    const breadcrumbEl = ref();
    const breadcrumbItemEl = ref([]);

    const handleClick = path => {
      router.push(path);
    };

    const shouldShowSeparator = key => {
      const { data } = props;

      return data.length - 1 !== key && data.length > 1;
    };

    const collectBreadcrumItemEl = el => {
      breadcrumbItemEl.value.push(el);
    };

    const changePage = path => {
      router.push(path);
    };

    return {
      breadcrumbEl,
      handleClick,
      shouldShowSeparator,
      collectBreadcrumItemEl,
      changePage,
      breadcrumbItemEl,
    };
  },
};
</script>
<style lang="scss" scoped>
.#{$prefixClass}-breadcrumb {
  @apply flex;

  &--item {
    @apply ml-2;

    &:last-child {
      @apply text-gray-400;
    }

    &:first-child {
      &::before {
        display: none;
      }
    }
  }

  .separator {
    @apply inline-block text-gray-400 px-2;
  }
}
</style>
