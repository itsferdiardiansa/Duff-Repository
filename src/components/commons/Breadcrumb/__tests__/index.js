import { mount } from '@vue/test-utils'
import Breadcrumb from '@common/Breadcrumb'
import Button from '@common/Button'
import { useRouter } from 'vue-router'

describe('common/Breadcrumb', () => {
  const router = useRouter()
  let Component
  let prefixClass

  beforeAll(async () => {
    Component = mount(Breadcrumb, {
      global: {
        stubs: {
          CMButton: Button,
        },
      },
    })
    prefixClass = `${Component.rootVM.prefixClass}-breadcrumb`
  })

  it('will render and match snapshot', () => {
    expect(Component.element).toMatchSnapshot()
  })

  it('will render the item according to the property data provided', async () => {
    const breadcrumbs = [
      { path: '/', name: 'Home' },
      { path: '/role', name: 'Role' },
      { path: '/role/create', name: 'Add Role' },
    ]

    await Component.setProps({ data: breadcrumbs })

    expect(Component.element).toMatchSnapshot()
  })

  it('it will trigger to click correctly', async () => {
    const breadcrumbs = [
      { path: '/', name: 'Home' },
      { path: '/partner', name: 'Partner' },
      { path: '/partner/create', name: 'Add Partner' },
    ]

    await Component.setProps({ data: breadcrumbs })

    await Component.vm.breadcrumbItemEl[1].querySelector('button').click()

    expect(router.push).toHaveBeenCalled()
    expect(router.push).toHaveBeenCalledWith('/partner')
  })
})
