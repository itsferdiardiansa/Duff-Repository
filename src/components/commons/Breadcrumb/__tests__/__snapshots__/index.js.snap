// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/Breadcrumb will render and match snapshot 1`] = `
<ul
  class="ss-breadcrumb"
>
  
  
</ul>
`;

exports[`common/Breadcrumb will render the item according to the property data provided 1`] = `
<ul
  class="ss-breadcrumb"
>
  
  <li
    class="ss-breadcrumb--item"
  >
    <button
      class="ss-button ss-button--primary ss-button--sm no-rounded"
    >
      <div
        class="ss-button--wrapper"
      >
        <!--v-if-->
        
        <div
          class="ss-button--label"
        >
          
          Home
          
        </div>
        
      </div>
    </button>
  </li>
  <li
    class="ss-breadcrumb--item"
  >
    <button
      class="ss-button ss-button--primary ss-button--sm no-rounded"
    >
      <div
        class="ss-button--wrapper"
      >
        <!--v-if-->
        
        <div
          class="ss-button--label"
        >
          
          Role
          
        </div>
        
      </div>
    </button>
  </li>
  <li
    class="ss-breadcrumb--item"
  >
    <button
      class="ss-button ss-button--primary-inverse ss-button--sm no-rounded disabled"
      disabled=""
    >
      <div
        class="ss-button--wrapper"
      >
        <!--v-if-->
        
        <div
          class="ss-button--label"
        >
          
          Add Role
          
        </div>
        
      </div>
    </button>
  </li>
  
</ul>
`;
